src/main/java/io/cdap/plugin/couchbase/CouchbaseConfig.java

```diff
@@ -0,0 +1,340 @@
+      Strings.isNullOrEmpty(user) && !Strings.isNullOrEmpty(password)) {
+      collector.addFailure("Username must be specified", null)
+        .withConfigProperty(CouchbaseConstants.USERNAME);
+    }
+    if (!containsMacro(CouchbaseConstants.USERNAME) && !containsMacro(CouchbaseConstants.PASSWORD) &&
+      !Strings.isNullOrEmpty(user) && Strings.isNullOrEmpty(password)) {
+      collector.addFailure("Password must be specified", null)
+        .withConfigProperty(CouchbaseConstants.PASSWORD);
+    }
+    if (!containsMacro(CouchbaseConstants.ON_ERROR)) {
+      if (Strings.isNullOrEmpty(onError)) {
+        collector.addFailure("Error handling must be specified", null)
+          .withConfigProperty(CouchbaseConstants.ON_ERROR);
+      } else if (ErrorHandling.fromDisplayName(onError) == null) {
+        collector.addFailure("Invalid record error handling strategy name",
+                             "Specify valid error handling strategy name")
+          .withConfigProperty(CouchbaseConstants.ON_ERROR);
+      }
+    }
+    if (!containsMacro(CouchbaseConstants.MAX_PARALLELISM) && maxParallelism == null) {
+      collector.addFailure("Max parallelism must be specified", null)
+        .withConfigProperty(CouchbaseConstants.MAX_PARALLELISM);
+    }
+    if (!containsMacro(CouchbaseConstants.SCAN_CONSISTENCY)) {
+      if (Strings.isNullOrEmpty(consistency)) {
+        collector.addFailure("Scan consistency must be specified", null)
+          .withConfigProperty(CouchbaseConstants.SCAN_CONSISTENCY);
+      } else if (Consistency.fromDisplayName(consistency) == null) {
+        collector.addFailure("Invalid scan consistency name", null)
+          .withConfigProperty(CouchbaseConstants.SCAN_CONSISTENCY);
+      }
+    }
+    if (!containsMacro(CouchbaseConstants.QUERY_TIMEOUT) && timeout == null) {
+      collector.addFailure("Query timeout must be specified", null)
+        .withConfigProperty(CouchbaseConstants.QUERY_TIMEOUT);
+    }
+    // TODO Couchbase Server 4.5 introduces INFER, a N1QL statement that infers the metadata of documents.
+    // This can be used to infer the Output Schema.
+    if (!containsMacro(CouchbaseConstants.SCHEMA) && Strings.isNullOrEmpty(schema)) {
+      collector.addFailure("Output schema must be specified", null)
+        .withConfigProperty(CouchbaseConstants.SCHEMA);
+    } else if (!containsMacro(CouchbaseConstants.SCHEMA)) {
+      Schema parsedSchema = getParsedSchema();
+      validateSchema(parsedSchema, collector);
+    }
+
+    collector.getOrThrowException();
+  }
+
+  private void validateSchema(Schema schema, FailureCollector collector) {
+    List<Schema.Field> fields = schema.getFields();
+    if (null == fields || fields.isEmpty()) {
+      collector.addFailure("Schema must contain at least one field", null)
+        .withConfigProperty(CouchbaseConstants.SCHEMA);
+      collector.getOrThrowException();
```

Comment:
I don't think it is necessary to throw exception here, we can simply return to skip fields validation.

Reply:
Changed.